### DISCUSSION TITLE
joint collision and cartesian collision

### DISCUSSION THREAD

**Discussion by matteolucchi** [justagist/panda_simulator#47] [Q&A]
Hi @justagist ! I hope you are doing well!

I am interested in the joint_collision and cartesian_collision values of the robot state. I have seen that they are included in the robot state also when using the panda simulator, I tried to manually send the robot into collision but I cannot manage to trigger any of the collision flags. I suppose that this behaviour is not implemented in simulation, is that correct or am I doing something wrong?

**Comment by justagist** [comment]
Hi @matteolucchi,
Yes, you are right, these flags are not triggered in the simulation.

**Reply by matteolucchi** [reply]
Ok thank you very much for the prompt reply!

**Reply by matteolucchi** [reply]
Regarding this, I have added Contact Sensors to the robot links to detect collisions and contacts with the robot itself and other objects in the simulation (https://github.com/jr-robotics/franka_panda_description/commit/19e8bd12438e5a7e5b01e142a4c0811f63a13b3d). The plugins publish the contact information on separate topics. If this is something that could be interesting for you to have in your repo let me know and I'd be happy to submit a pull request.

**Reply by justagist** [reply]
Hi @matteolucchi, please do submit a PR with your changes; it could be very useful, thank you!

By briefly going through your commit, I see that your changes would attach contact/bump sensors on each link for detecting contacts. I agree this would be a simple solution for reliably triggering the cartesian flags that you mentioned (obviously this is not how it is done on the real Franka robot, where the cartesian contact forces are estimated from joint torques; but your solution is much simpler and should be enough for the purposes of the simulator for now).
I assume the plugin publishes some measure of cartesian force+torque (or maybe just force?) for a contact, and for each link in a separate topic? If so, we should be able to use these values to trigger the 'cartesian_collision' and 'cartesian_contact' values in the robot state by comparing them against two user-defined sets of thresholds.
I wanted to know if you have any ideas/advice on how we can similarly trigger the 'joint_contact' and 'joint_collisions' values? These are based on joint torque values as far as I understand. We can obtain measurements of the commanded and actual values of the joint torques from the simulator. I think that the joint contact/collision thresholds are based on the difference between the commanded/expected and the actual joint torque values, but I am not sure. I would love to hear your thoughts on this if you don't mind.

**Reply by matteolucchi** [reply]
Yes you are right, the plugin publishes [ContactsState message](http://docs.ros.org/en/noetic/api/gazebo_msgs/html/msg/ContactsState.html) on a separate topic for each link, this includes the name of the contact bodies and forces and torques. I currently do not use the full information of this message, in my case I just check if there is any contact and in that case I assume there was either a self collision or a collision with the environment. 

My doubt is whether it is better to integrate this to trigger the default flags of the panda interface, or given that this works with a different concept, to simply provide this extra information published via ros topic as it is already done now. What are your thoughts on that? 

Regarding the joint_contacts and joint_collision unfortunately I do not know more than you on how these work on the real robot, I might look into this in the future and if I come up with any idea on how to implement these in the simulated version I'll definitely report back here. I do think that it works like you imagine, and your suggested approach for implementing this in simulation also seems like a good idea, my only doubt here is how reliable the joint torque measurement is. Again, if I have time I might try to do a comparison between the values on the real robot and on the simulated one to see if the torque values are somehow close.

**Reply by justagist** [reply]
> My doubt is whether it is better to integrate this to trigger the default flags of the panda interface, or given that this works with a different concept, to simply provide this extra information published via ros topic as it is already done now. What are your thoughts on that?

I am not particularly keen on trying to imitate the mechanisms and implementations from the real robot as this could be really hard, especially since many of the internal implementations of the Franka control is not open-source. I am more interested in simulating a similar behavioural response from the simulated robot when using the Franka ROS Interface API calls, which would help in reusing the same code for the simulator and real robot (with minimum change -- ideally only in parameters such as controller gains, pre-set thresholds, etc.). In fact, I use a force-torque sensor plugin at the end-effector to compute end-effector wrench in the simulator, unlike the real robot which does not have a separate end-effector FT sensor (and is computed internally using joint torques). So I propose we can use the information from the plugin as a proxy for detecting collisions and as a measure for contact forces (at least until we find that it causes some undesired effects).

> my only doubt here is how reliable the joint torque measurement is

That is a good point. I too am not sure how reliable the absolute values will be (given that this will be massively influenced by the dynamics parameters set in the description files of the robot), but I would guess that the error values (difference between expected and computed) will be reasonably reliable..? I guess this would depend on the implementation of the joint controller as well...

> Again, if I have time I might try to do a comparison between the values on the real robot and on the simulated one to see if the torque values are somehow close.

That would mainly depend on the accuracy of the dynamics model in the robot description, I believe, and may not be very useful for this particular issue. If I get a chance, I will do some experiments to see how the *difference* b/w commanded and measured torques are affected due to external forces/contacts.

> I might look into this in the future and if I come up with any idea on how to implement these in the simulated version I'll definitely report back here.

Thanks a lot @matteolucchi! I will work on this as well when I get a chance, and will let you know if I find anything useful.

**Reply by matteolucchi** [reply]
> I am not particularly keen on trying to imitate the mechanisms and implementations from the real robot as this could be really hard, especially since many of the internal implementations of the Franka control is not open-source. I am more interested in simulating a similar behavioural response from the simulated robot when using the Franka ROS Interface API calls, which would help in reusing the same code for the simulator and real robot (with minimum change -- ideally only in parameters such as controller gains, pre-set thresholds, etc.). In fact, I use a force-torque sensor plugin at the end-effector to compute end-effector wrench in the simulator, unlike the real robot which does not have a separate end-effector FT sensor (and is computed internally using joint torques). So I propose we can use the information from the plugin as a proxy for detecting collisions and as a measure for contact forces (at least until we find that it causes some undesired effects).

Ok then I will open a pull request with what I have so far and you can let me know if this is ready to be merged. 


Sounds good, I will update you if I get to something useful!

**Reply by justagist** [reply]
I have merged your changes to the description package: https://github.com/justagist/franka_panda_description/pull/6.
I will look into using this to update the robot_state and the collision/contact flags in the simulator when I get a chance.
Thanks again!

**Reply by matteolucchi** [reply]
Thank you! And thank you again for the great work!
